### PR TITLE
RELATED: FET-608 - Finalize pivot table refactoring

### PR DESCRIPTION
**Pull request by lupko**
-  Rename table column descriptors
-  Mild refactoring of the col descriptor types
-  Documentation improvements
-  Loading renderer clarification & improvements
-  Testing/mocking infrastructure for table facade
-  Lifecycle optimizations
   -  Table now explicitly covers state when its state is transforming due to change of sorts or totals
   -  Requests to re-render will check against the transformation target state, preventing full reinits in AD/KD
-  Performance optimizations
   -  Leak in caching abstraction was causing table execution transformation to avoid cache

Related:

-  Fixing tsconfig.json for pivot and ui-kit. This was causing weird TS errors in test code
-  Fix ui-kit build so that api-extractor is only run once
-  Fix ui-kit build so that dev mode runs in foreground

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2744/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4377/ : SUCCESS in 15m 27s

**Comment by lupko** [comment]
extended test

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/722/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2746/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/722/ : UNSTABLE in 32m 21s

**Comment by lupko** [comment]
extended test

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/723/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2748/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/723/ : SUCCESS in 21m 03s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2758/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4386/ : SUCCESS in 15m 39s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2759/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4387/ : SUCCESS in 15m 42s

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx [hunk 1]
```diff
@@ -314,18 +342,46 @@ export class CorePivotTableAgImpl extends React.Component<ICorePivotTableProps,
...
         if (!this.internal.table) {
-            // table is not yet initialized. do reinit in case the execution being initialized (one in prevProps) is
-            // different from the new execution.
-            return !prepExecutionSame;
+            // Table is not yet fully initialized. See if the initialization is in progress. If so, see if
+            // the init is for same execution or not. Otherwise fall back to compare props vs props.
+            //
+            // Note: testing the exec used by initializer is crucial in contexts (say AD) where for instance
+            // a change in sorts triggers new reinit in table itself and also sends an event up to the app. The app
+            // then sends the
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/CorePivotTable.tsx:357]: Missing end of comment.

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx [hunk 2]
```diff
@@ -570,13 +629,34 @@ export class CorePivotTableAgImpl extends React.Component<ICorePivotTableProps,
...
+     * LoadingRenderer used in all cells of the left-most column.
+     *
+     * The table must take care to remove the sticky (top-pinned) row - it is treated differently by
+     * ag-grid and will be literally sticking there on its own with the loading indicators.
+     *
+     * Once transformation finishes - indicated by call to onPageLoaded, table can re-instance the sticky row.
+     *
+     * @param _newExecution
+     */
+    private onExecutionTransformed = (_newExecution: IPreparedExecution): void => {
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/CorePivotTable.tsx:648]: Do we need the parameter anymore?

File: libs/sdk-ui-pivot/src/impl/data/loadingRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * indicators as data for them is being loaded.
+ *
+ * NOTE: keep in mind that this loading renderer IS NOT used to configure ag-grids built-in loading renderer. This
+ * renderer is essentially a custom cell renderer which our table impl uses for all cells of the left-most table column.
+ *
+ * @param _table
+ * @param _props
+ */
+export function createLoadingRenderer(
+    _table: TableFacade,
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/data/loadingRenderer.tsx:20]: Why do you need the factory? You are not "baking in" anything, the parameters of the factory are not used.

Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/loadingRenderer.tsx:20]: this is here for consistency with other cell renderers which do need to access table / table descriptor. the loading renderer should be just a decorator that adds the indicator on top the normal cell renderer.

one refactor ahead, if we want to use renderer composition (loading indicator on top of the standard cell renderer) we will need these parameters.

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx (hunk 1 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/CorePivotTable.tsx:357]: i've cleaned it up

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx (hunk 2 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/CorePivotTable.tsx:648]: nope, we don't. i thought keeping it here as unused may be clearer IF we need it in the future.

no strong opinions i can remove it.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2765/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4392/ : SUCCESS in 18m 25s

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx (hunk 2 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/CorePivotTable.tsx:644]: Fair, leave it in :)

File: libs/sdk-ui-pivot/src/impl/data/loadingRenderer.tsx (hunk 3 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/impl/data/loadingRenderer.tsx:20]: Fair, leave it in :)

**Comment by lupko** [comment]
This is OK to merge. The code is finalized. There are couple of big things flying around now so want to wait for a slightly 'calmer' time to push these optimizations.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2879/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4509/ : SUCCESS in 21m 44s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3268/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4927/ : FAILURE in 5m 50s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3270/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4929/ : SUCCESS in 12m 17s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3274/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4933/ : SUCCESS in 12m 28s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result
 - No new bug found. Congratulation!

More detail, see in https://sonarqube-dev.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1110

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/987/ : SUCCESS in 3m 14s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1257/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4936/ : SUCCESS in 20m 51s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-el7/375/ : SUCCESS in 8m 57s (non-voting)

**Pull request merged**